provider/pkg/resources/extension_resource.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2024, Pulumi Corporation.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/mattbaird/jsonpatch"
+	"github.com/pulumi/pulumi-aws-native/provider/pkg/client"
+	"github.com/pulumi/pulumi-aws-native/provider/pkg/default_tags"
+	"github.com/pulumi/pulumi-go-provider/resourcex"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/wI2L/jsondiff"
+)
+
+type ExtensionResource struct {
+	Type         string
+	Properties   map[string]any
+	WriteOnly    []string
+	TagsProperty string
+	TagsStyle    default_tags.TagsStyle
+}
+
+type extensionResource struct {
+	client client.CloudControlClient
+}
+
+func NewExtensionResource(client client.CloudControlClient) *extensionResource {
+	return &extensionResource{
+		client: client,
+	}
+}
+
+func (r *extensionResource) Check(ctx context.Context, urn resource.URN, inputs, state resource.PropertyMap, defaultTags map[string]string) (resource.PropertyMap, []ValidationFailure, error) {
+	var extensionResource ExtensionResource
```

Comment:
nit. I find the variable name and two type names all being [Ee]xtensionResource to be kind of confusing to read.